ignite-shuffle-spark-io/src/main/java/com/palantir/ignite/spark/shuffle/io/IgniteBlockFetcherIterator.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.ignite.spark.shuffle.io;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.palantir.ignite.SparkShufflePartition;
+import com.palantir.ignite.SparkShufflePartitionBlock;
+import com.palantir.logsafe.SafeArg;
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.io.SequenceInputStream;
+import java.util.Deque;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.LongStream;
+import java.util.stream.StreamSupport;
+import org.apache.ignite.IgniteCache;
+import org.apache.ignite.lang.IgniteFuture;
+import org.apache.spark.TaskContext;
+import org.apache.spark.api.shuffle.ShuffleBlockInfo;
+import org.apache.spark.api.shuffle.ShuffleReadSupport;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class IgniteBlockFetcherIterator implements ShuffleReadSupport.ShuffleReaderIterator {
+
+    private static final Logger log = LoggerFactory.getLogger(IgniteBlockFetcherIterator.class);
+
+    private final String appId;
+    private final Map<SparkShufflePartitionBlock, byte[]> currentBlocks = Maps.newHashMap();
+    private final LinkedBlockingDeque<IgniteFuture<Map<SparkShufflePartitionBlock, byte[]>>> runningTasks =
```

Comment:
@squito @yifeih so this is an example of how we could implement the shuffle reader APIs from https://github.com/palantir/spark/pull/523 using a different storage backend, in this case Apache Ignite. I think this branch is out of date though so I need to publish Yifei's latest iterations to my local maven repository to work against it.

Here's some of the highlights I would note from this thought experiment:

- Pipelining can be implemented behind the API, not in front of it. I think the notion of parallel fetches and whether or not to fetch to disk vs. buffer blocks in memory should be left up to the plugin to decide.
- Stream corruption should more or less be impossible given the guarantees of Ignite as a system. I wonder if we care about stream corruption in any case other than the cases where we're using Netty as the transfer protocol. We could put stream corruption checks as a special case, e.g. we check if the shuffle reader is of type `DefaultShuffleBlockReader` or `ShuffleBlockFetcherIterator` and have an additional check layer, etc.
- We rely on Ignite to be fully reliable - there isn't any retry logic here for any block fetches.

Reply:
> Pipelining can be implemented behind the API, not in front of it. I think the notion of parallel fetches and whether or not to fetch to disk vs. buffer blocks in memory should be left up to the plugin to decide.

I agree that it is fine to leave that up to the plugin.  (I can't think of a case where you wouldn't want pipelining at the moment, but there is probably some case ...)  I do think it would be nice if there were some reusable code to allow plugins to get pipelining, without having to implement it themselves from scratch.

I feel a little more strongly that fetch-to-disk should *not* be part of the plugin, since it has more to do with spark's memory management, but again its fine if its behind the api, just would be nice if there is reusable code for it.

> We rely on Ignite to be fully reliable - there isn't any retry logic here for any block fetches.

I'm not certain I know what this means.  (caveat: I don't really know anything about ignite, I'm viewing it as having similar properties to hdfs for these purposes.)  Its certainly possible that an executor fails to read data from ignite: something is wrong with the executor, or the entire ignite cluster is unreachable or overloaded, or somebody accidentally wipes data from ignite.

I guess you are saying you have no explicit retries within one task from the ignite plugin, nor any special failure handling.  But does that mean a failure is treated as a regular exception, so the task gets retried 4 times and then the job is failed?  That means that a failure to read from ignite is catastrophic, and the spark app cannot recover -- that may be fine given your expectations from ignite.  Or will it get translated to a FetchFailedException (by some intermediate parts of spark's code which surrounds this plugin), which will lead the scheduler to mark some data as missing and regenerate it?  If so, are we sure its marking the right data as missing?

Reply:
> I feel a little more strongly that fetch-to-disk should not be part of the plugin, since it has more to do with spark's memory management, but again its fine if its behind the api, just would be nice if there is reusable code for it.

The problem is that the benefits of fetch to disk is that the data doesn't have to be copied into memory at all. But I think for blob stores that don't support network-to-disk direct download that this is more or less impossible. For example in Ignite we'd have to get the `byte[]` in memory at some point, so here we say that we don't want to both transfer into memory and also transfer to disk. We might as well keep all the data in one of three places:
- Memory of the executor
- Memory or disk of the Ignite store (that's up to Ignite's implementation and configuration)
- Over the network

> I guess you are saying you have no explicit retries within one task from the ignite plugin, nor any special failure handling. But does that mean a failure is treated as a regular exception, so the task gets retried 4 times and then the job is failed? That means that a failure to read from ignite is catastrophic, and the spark app cannot recover -- that may be fine given your expectations from ignite.

Yup this is the case - failing to read from Ignite should be such a rare case that it's probably fine to assume the job is not recoverable if the cluster's blocks can't be fetched that many times. But I also have a lot to learn about the specific QoS guarantees of Ignite. Can we have similar expectations for HDFS? We're thinking about trying to implement this plugin as an experiment against S3 - I'd expect S3 to have similar kinds of guarantees, especially since S3 is also a persistent data store for many Spark applications. So for example, failing to read from S3 as a source dataset for a data frame would certainly cause a job failure anyways.